src/Microsoft.ML.Data/Model/Onnx/ICanSaveOnnx.cs

```diff
@@ -19,9 +19,9 @@ public interface ICanSaveOnnx
     }
 
     /// <summary>
-    /// This data model component is savable as ONNX.
+    /// This component know how to save himself in ONNX format.
     /// </summary>
-    public interface ITransformCanSaveOnnx: ICanSaveOnnx, IDataTransform
+    public interface ISaveAsOnnx
```

Comment:
>ISaveAsOnnx [](start = 21, length = 11)

This thing needs to descend from `ICanSaveOnnx`, because it is often the case, as you'll see if you inspect the actual implementations of `CanSaveOnnx`, that the behavior of whether or not we can in fact save as ONNX is far more complex than whether we implement the interface or not.

Reply:
For example, if there were an implementation of `IRowMapper` that wrapped any existing mapper, obviously that meta-wrapper would have to report it can save as ONNX, but internally any of the things it is wrapping do not it is hosed, which is bad.

By having the core interface always have this runtime check, we avoid the problem that we have elsewhere with things like exporting as text, exporting as INI, and so on, where we have things that report they *can* save as something, but really cannot, and fail in strange and interesting ways.

---
In reply to: [214432408](https://github.com/dotnet/machinelearning/pull/759#discussion_r214432408) [](ancestors = 214432408)